### PR TITLE
Add manager support

### PR DESCRIPTION
**Pull request by stakach**
Allows you to create a simple service for accepting traps - also supports multiple engine_ids / communities.

What do you think? I haven't included tests yet but happy to do so if you would consider accepting this pull request.

Also feel free to make any changes, I am not a SNMP expert, just need to support traps for monitoring switches (as switch interfaces go high and low)

Example usage:

```ruby
require 'libuv'
require 'netsnmp/manager'

class Manager < NETSNMP::Manager
  def initialize(port)
    super(*[])

    @transport = ::Libuv.reactor.udp { |data, ip, port|
      new_message(data, ip, port)
    }.bind('0.0.0.0', port).start_read
  end

  def send(ip, port, data)
    @transport.send(ip, port, data)
  end
end

class MyHandler
  def inform(request_pdu, ip, port, manager)
    puts "handler got inform: #{request_pdu}"
  end

  def security_parameters
    {
      security_level: :authpriv, username: "mustermann", 
      auth_protocol: :md5, priv_protocol: :aes,
      auth_password: "authpass", priv_password: "privpass",
      engine_id: "\x00\x00\xDE\xAD\x00\x00\xBE\xEF"
    }
  end
end

reactor do |reactor|
  man = Manager.new(162)
  man.add_community(MyHandler.new, name: "public")
end

```

Then it will process trap and inform requests (example send inform request)

```ruby
require 'netsnmp'

manager = NETSNMP::Client.new(host: "localhost", port: 162,
                                                username: "mustermann", security_level: :authpriv,
                                                auth_protocol: :md5, priv_protocol: :aes,
                                                engine_id: "\x00\x00\xDE\xAD\x00\x00\xBE\xEF", # as set in manager
                                                auth_password: "authpass", priv_password: "privpass")

manager.inform oid: "1.3.6.1.2.1.1.3.0", value: 43

```

### CODE REVIEW

**Comment by HoneyryderChuck** [comment]
@stakach first of all, thank you for the PR! Indeed, trap support has been lacking for some time.

i haven't been getting many PRs, and don't know how high the entry barrier is, so feel free to giving me inputs about the code structure or how easy it is for you to test (or whether you get how things are tested).

I had a quick look at you code, and this is what I would do:

* Manager seems to be a specialized Client (Client is a Manager? Never got the nomenclature right here).
* I think that some of code should rewrite the Session, particularly the logic for receiving PDUs and sending a response
  * `Session` implements `#send`, which might not be the best name for what it does; Ideally, `#send`would just send, `#recv` would just receive; If you need to rename these specific methods, please do;
  * The Manager should just be able to receive (implements `#recv`);
* Trap and Inform should be specialized first-class PDUs, just like [the already existing](https://github.com/swisscom/ruby-netsnmp/blob/master/lib/netsnmp/pdu.rb#L47-L58).

Of course, some tests will be needed. But this is really what I think that the scope of changes should be. I don't think that the handler belongs here. Handler should be user code, the library only knows how to receive and decode trap PDUs. 

(This comment is a work in progress)

**Comment by HoneyryderChuck** [comment]
Continuing on the previous comment about the handler:

I think the Manager (or Client) should have a method (`#receive`, `#receive_pdu`, `#trap`...) just expecting to receive a PDU. And the user should handle the PDU:

```ruby
until descriptor_is_ready?
  pdu = manager.trap
  # handle pdu
  if is_trap?(pdu)
     ....
end
```

what do you think?

**Comment by stakach** [comment]
Definitely agree about the first-class PDUs. This would make it easier to send traps and informs.

The reason I made manager its own thing as it plays a slightly different role
* Agent (port 161) - accepts get and set requests
* Manager (port 162) - accepts trap and inform requests
* Client - talks to Agents to request information and potentially sends traps / informs to managers

If we want to keep this library focused on being the best client implementation possible then maybe we just add the additional PDUs for sending traps / informs.

However as clients are often also managers I thought it would be nice to offer a base class that could be used to implement a manager. Even if it is more of an example of what one should do to implement a manager - which is why I cleanly separated its logic from the rest of the library.

Basically the whole SNMP spec is pretty convoluted, but I think you've done a great job of making it look as simple as possible

**Comment by HoneyryderChuck** [comment]
@stakach thx for the kind words.

I mentioned merging Manager and Client because, at least according to RFC and general SNMP nomenclature used, there is no client, only managers and agents :) . I don't feel strongly opposed to separate the concepts for now, though.

However, I'd keep the same structure: Client and Manager should be the same internally (aka: hold a session), just differ in the public API used.

My suggestion still holds: Manager should have a receive method (I tried to look in other implementations for a better name, but haven't found one better than `#receive_pdu`, open to suggestions). This method should be perceived as synchronous:

```ruby
# baby implementation
loop do
  pdu = manager.receive_pdu
  case pdu
    ...
end
```

As suggested, some rewording of the session API is open to discussion (`Session#send` should only send, not receive as well). Handler should be presented more as a pattern than an implementation detail. 

Can I suggest you to start with writing the specs, in order for the API to be defined? This will also tell me whether you find the whole test suite hard to work with, and I can help you navigate it or even make it easier if need be.

**Comment by HoneyryderChuck** [comment]
I read again your comment, and I may have read it too quick.

Is the purpose to create a trap listener, or a trap sender (or both)? Because typically the Manager/Client receives traps/informs, and the agent send traps/informs. 

This library has been until now mostly a client library. Implementing the trap sender rule would be venturing into agent territory, for which I'd say there is not a lot of building blocks. 

If my comments have confused you, it's because I am mostly thinking about receiving traps. Apologies for that.

**Comment by stakach** [comment]
There is no reason that this library couldn't be used to implement an Agent i.e. be capable of sending informs or traps - hence why I added the `inform` function

**Comment by HoneyryderChuck** [comment]
I agree, but that should be then under an `Agent` entity, it falls out of the Manager scope.

So your purpose is to implement an agent and keep sending trap/inform PDUs? Or the manager functionality (trap listener) is also wanted?

**Comment by HoneyryderChuck** [comment]
@stakach I don't know what happened here, but I changed a lot of info from the commit tree (regarding my email), and it seems to have closed all active PRs(?). And github doesn't let me reopen :/

I also published a new version (0.1.4), and added rubocop to the mix. Would you consider migrating your changes to a new PR? I don't think I will be doing any more commit-level changes.
